### PR TITLE
BI-8388 Implement order notification sender

### PR DESCRIPTION
**Pull request by dlgroves-ch**
* Incoming message deserialised by MessageDeserialiser into an appropriate entity.
* Message enriched with an OrderApi resource fetched with OrdersApiService and mapped to an OrderModel instance using an OrdersApiMapper appropriate to the given item kind.
* Resulting message is serialised by MessageFactory into Avro and sent to the email-send Kafka topic by KafkaProducer.
* If enrichment fails then the message will be put onto the order-received-notification-retry topic and re-attempted. 
* A message will subsequently be published to the order-received-notification-error topic if the configured number of retries is exhausted.

Resolves:
[BI-8388](https://companieshouse.atlassian.net/browse/BI-8388)

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/ordernotification/emailmodel/OrderResourceOrderNotificationEnricher.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        this.loggingUtils = loggingUtils;
+    }
+
+    /**
+     * Enriches an order received notification with an order resource fetched from the Orders API.
+     *
+     * @param orderReference the order responsible for triggering the notification
+     */
+    public EmailSend enrich(final String orderReference) throws OrdersResponseException {
+        final OrdersApi order;
```
Comment (kinpang-CH) [inline src/main/java/uk/gov/companieshouse/ordernotification/emailmodel/OrderResourceOrderNotificationEnricher.java]: Remove line and use declaration at 49

File: src/main/java/uk/gov/companieshouse/ordernotification/orders/service/OrdersApiService.java [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package uk.gov.companieshouse.ordernotification.orders.service;
+
+import uk.gov.companieshouse.api.model.order.OrdersApi;
+
+public interface OrdersApiService {
```
Comment (kinpang-CH) [inline src/main/java/uk/gov/companieshouse/ordernotification/orders/service/OrdersApiService.java]: Change interface name to OrderRetrievable

File: src/main/java/uk/gov/companieshouse/ordernotification/orders/service/OrdersApiServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+@Service
+class OrdersApiServiceImpl implements OrdersApiService {
+    
+    private static final Logger LOGGER = LoggerFactory.getLogger(LoggingUtils.APPLICATION_NAMESPACE);
+    
+    private final ApiClient apiClient;
+    private final LoggingUtils loggingUtils;
+
+    @Autowired
+    public OrdersApiServiceImpl(ApiClient apiClient, LoggingUtils loggingUtils) {
```
Comment (kinpang-CH) [inline src/main/java/uk/gov/companieshouse/ordernotification/orders/service/OrdersApiServiceImpl.java]: Rename class to OrdersApiOrderRetriever

File: src/test/java/uk/gov/companieshouse/ordernotification/messageproducer/MessageFactoryTest.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+    private static final String APP_ID = "App Id";
+    private static final String EMAIL_DATA = "Message content";
+    private static final String EMAIL_ADDR = "someone@example.com";
+    private static final String MSG_ID = "Message Id";
+    private static final String MSG_TYPE = "Message type";
+    private static final String ORDER_REF = "ORD-ABC-123";
+    private static final String CREATED_AT = "2020-08-25T09:27:09.519+01:00";
+
+    private static final String TOPIC = "email-send";
```
Comment (kinpang-CH) [inline src/test/java/uk/gov/companieshouse/ordernotification/messageproducer/MessageFactoryTest.java]: Place above constants inside TestConstants.java

File: src/test/java/uk/gov/companieshouse/ordernotification/orders/service/OrdersApiServiceImplTest.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+
+@ExtendWith(MockitoExtension.class)
+public class OrdersApiServiceImplTest {
```
Comment (kinpang-CH) [inline src/test/java/uk/gov/companieshouse/ordernotification/orders/service/OrdersApiServiceImplTest.java]: Rename test class OrdersApiOrderRetrieverTest

File: src/test/java/uk/gov/companieshouse/ordernotification/ordersconsumer/OrderProcessingException.java [hunk 6]
```diff
@@ -0,0 +1,5 @@
+package uk.gov.companieshouse.ordernotification.ordersconsumer;
+
+public class OrderProcessingException extends RuntimeException {
+    public OrderProcessingException() { super("Order processing failed."); }
+}
```
Comment (kinpang-CH) [inline src/test/java/uk/gov/companieshouse/ordernotification/ordersconsumer/OrderProcessingException.java]: Remove this class

File: src/test/java/uk/gov/companieshouse/ordernotification/ordersconsumer/OrdersKafkaConsumerTest.java [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+class OrdersKafkaConsumerTest {
+    private static final String ORDER_RECEIVED_URI = "/order/ORDER-12345";
+    private static final String ORDER_RECEIVED_TOPIC = "order-received";
+    private static final String ORDER_RECEIVED_KEY = "order-received";
+    private static final String ORDER_RECEIVED_TOPIC_RETRY = "order-received-notification-retry";
+    private static final String ORDER_RECEIVED_TOPIC_ERROR = "order-received-notification-error";
+
```
Comment (kinpang-CH) [inline src/test/java/uk/gov/companieshouse/ordernotification/ordersconsumer/OrdersKafkaConsumerTest.java:37]: Extract constants to TestConstants.java

File: src/test/resources/application-stubbed.properties [hunk 8]
```diff
@@ -0,0 +1,6 @@
+spring.kafka.bootstrap-servers = kafka-bootstrap:9092
+uk.gov.companieshouse.item-handler.error-consumer = false
+
+email.senderAddress = noreply@companieshouse.gov.uk
+email.producerTimeout = 30
+maximum.retries = 3
```
Comment (kinpang-CH) [inline src/test/resources/application-stubbed.properties:6]: Can we get away with using just this properties file?

**Review by kinpang-CH — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/companieshouse/ordernotification/emailmodel/OrderResourceOrderNotificationEnricher.java (hunk 1 above)
Comment (cgregoryCH) [reply src/main/java/uk/gov/companieshouse/ordernotification/emailmodel/OrderResourceOrderNotificationEnricher.java]: Resolved in latest commit

File: src/main/java/uk/gov/companieshouse/ordernotification/orders/service/OrdersApiService.java (hunk 2 above)
Comment (cgregoryCH) [reply src/main/java/uk/gov/companieshouse/ordernotification/orders/service/OrdersApiService.java]: Resolved in latest commit

File: src/main/java/uk/gov/companieshouse/ordernotification/orders/service/OrdersApiServiceImpl.java (hunk 3 above)
Comment (cgregoryCH) [reply src/main/java/uk/gov/companieshouse/ordernotification/orders/service/OrdersApiServiceImpl.java]: Resolved in latest commit

File: src/test/java/uk/gov/companieshouse/ordernotification/messageproducer/MessageFactoryTest.java (hunk 4 above)
Comment (cgregoryCH) [reply src/test/java/uk/gov/companieshouse/ordernotification/messageproducer/MessageFactoryTest.java]: Resolved in latest commit

File: src/test/java/uk/gov/companieshouse/ordernotification/orders/service/OrdersApiServiceImplTest.java (hunk 5 above)
Comment (cgregoryCH) [reply src/test/java/uk/gov/companieshouse/ordernotification/orders/service/OrdersApiServiceImplTest.java]: Resolved in latest commit

File: src/test/java/uk/gov/companieshouse/ordernotification/ordersconsumer/OrderProcessingException.java (hunk 6 above)
Comment (cgregoryCH) [reply src/test/java/uk/gov/companieshouse/ordernotification/ordersconsumer/OrderProcessingException.java]: Resolved in latest commit

File: src/test/java/uk/gov/companieshouse/ordernotification/ordersconsumer/OrdersKafkaConsumerTest.java (hunk 7 above)
Comment (cgregoryCH) [reply src/test/java/uk/gov/companieshouse/ordernotification/ordersconsumer/OrdersKafkaConsumerTest.java:37]: Resolved in latest commit

File: src/test/resources/application-stubbed.properties (hunk 8 above)
Comment (cgregoryCH) [reply src/test/resources/application-stubbed.properties:6]: Unfortunately not

**Review by cgregoryCH — APPROVED** [review]

**Review by kinpang-CH — APPROVED** [review]

**Review by cgregoryCH — APPROVED** [review]

**Pull request merged**